### PR TITLE
Redesign the DocNode hierarchy to support manual creation

### PR DESCRIPTION
**Pull request by pgonzal**
The first draft of the `DocNode` design was focused around accurately capturing the text ranges that the parser used to build the tree.  However, we also have a scenario where people will use the API to build up `DocNode` trees to be rendered as TSDoc comments, in which case these text ranges will not be relevant.

This PR decouples the text range information into an `Excerpt` class that is optionally associated with each `DocNode` object.  The `ParserContext` is now a public API and is the return value of `TSDocParser.parseRange()`.  Each `Excerpt` objects has a reference to the associated `ParserContext` object.  In a manual creation scenario, the `Excerpt` and `ParserContext` objects are not relevant and do not need to be constructed at all.

This PR also lays the necessary groundwork for the (upcoming) later parser stage that will partition the `DocNode` tree into well-known blocks for `@summary`, `@remarks`, `@params`, and `@returns`.

### CODE REVIEW

File: tsdoc/src/nodes/DocErrorText.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  public readonly errorMessage: string;
+
+  /**
+   * The range of characters that caused the error.  In general these may be
+   * somewhat farther ahead in the input stream from the DocErrorText node itself.
+   *
+   * @remarks
+   * For example, for the malformed HTML tag `<a href="123 /a>`, the DocErrorText node
+   * will correspond to the `<` character that looked like an HTML tag, whereas the
+   * error location might be the `"` character that was missing its closing pair.
```
Comment (iclanton) [inline tsdoc/src/nodes/DocErrorText.ts]: This example isn't great. Maybe `<a href="asdf" @ />`

File: tsdoc/src/parser/Excerpt.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  /**
+   * The ParserContext object from the parser invocation that processed the input.
+   */
+  public readonly parserContext: ParserContext;
+
+  /**
+   * The main textual content for the associated node.
+   */
+  public readonly prefix: TokenRange;
+  /**
```
Comment (iclanton) [inline tsdoc/src/parser/Excerpt.ts]: >/** [](start = 2, length = 3)

newline.

File: tsdoc/src/parser/TokenRange.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    this.parserContext = parameters.parserContext;
+    this._pos = parameters.pos;
+    this._end = parameters.end;
+    this._validateBounds();
+  }
+
+  /**
+   * The starting index into the associated `ParserContext.tokens` list.
+   */
+  public get pos(): number {
```
Comment (iclanton) [inline tsdoc/src/parser/TokenRange.ts]: >pos [](start = 13, length = 3)

Using the same terminology the TS compiler uses to refer to text locations is a little confusing here.

Perhaps call this "tokenStartIndex"/"tokenEndIndex"

File: tsdoc/src/parser/TokenReader.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  private _currentIndex: number;
+
+  public constructor(parserContext: ParserContext) {
+    this._parserContext = parserContext;
+    this.tokens = parserContext.tokens;
+    this._rangeStartIndex = 0;
+    this._currentIndex = 0;
+  }
+
+  public extractQueue(): TokenRange {
```
Comment (iclanton) [inline tsdoc/src/parser/TokenReader.ts]: >Queue [](start = 16, length = 5)

Field?

Comment (iclanton) [reply tsdoc/src/parser/TokenReader.ts]: Sequence?

---
In reply to: [209787204](https://github.com/Microsoft/tsdoc/pull/39#discussion_r209787204) [](ancestors = 209787204)

Comment (pgonzal) [reply tsdoc/src/parser/TokenReader.ts]: I renamed the class to `TokenSequence`.  And I renamed `extractQueue()` to `extractAccumulatedSequence()`

---
In reply to: [209787255](https://github.com/Microsoft/tsdoc/pull/39#discussion_r209787255) [](ancestors = 209787255,209787204)

**Pull request merged**